docs/console/incremental-guide.md

```diff
@@ -0,0 +1,586 @@
+           {
+               Console.WriteLine("Hello World!");
+           }
+       }
+   }
+   ```
+
+2. `$ dotnet restore`
+
+   `dotnet restore` analyzed the `project.json` file, downloaded the dependencies stated in the file (or grabbed them from a cache on your machine), and wrote the `project.lock.json` file.  The `project.lock.json` file is necessary to be able to compile and run.
+   
+   The `project.lock.json` file is a persisted and complete set of NuGet dependencies and other information describing an app.  This file is read by other tools, such as `dotnet compile` and `dotnet run`, enabling them to process the source code with a correct set of NuGet dependencies and binding resolutions.
+   
+3. `$ dotnet run`
+
+   `dotnet compile` compiled the source in Intermediate Language (IL) and generated an executable shim and a `Hello.dll` `.dll` file which contains the IL.
+   
+You can now invoke the executable.
+
+```
+$ /bin/Debug/dnxcore50/Hello.exe
+Hello, World!
+```
+
+### Other ways to compile
+
+Let's try compiling `dotnet run` to compile and execute the code without generating any build artifacts.  Note that although this has use (mainly by ASP.NET internally), it's not a recommended way to run build and run console applications.
+
+```
+$ dotnet run
+Hello, World!
+```
+
+Let's try compiling a native binary instead.
+
+```
+$ dotnet compile --native
+$ ./bin/Debug/dnxcore50/native/Basic
+Hello World!
+```
+
+Note the difference in compile time versus execution time.
+
+### Augmenting the program
+
+Let's change the file just a little bit.  Fibonacci numbers are fun, so let's try that out:
+
+`Program.cs`:
+
+```csharp
+using System;
+
+namespace ConsoleApplication
+{
+    public class Program
+    {
+        public static int FibonacciNumber(int n)
+        {
+            int a = 0;
+            int b = 1;
+            int tmp;
+            
+            for (int i = 0; i < n; i++)
+            {
+                tmp = a;
+                a = b;
+                b += tmp;
+            }
+            
+            return a;   
+        }
+        
+        public static void Main(string[] args)
+        {
+            Console.WriteLine("Hello World!");
+            Console.WriteLine("Fibonacci Numbers 1-15:");
+            
+            for (int i = 0; i < 15; i++)
+            {
+                Console.WriteLine($"{i+1}: {FibonacciNumber(i)}");
+            }
+        }
+    }
+}
+```
+
+And running the program:
+
+```
+$ dotnet compile --native
+$ ./bin/Debug/dnxcore50/native/Basic
+1: 0
+2: 1
+3: 1
+4: 2
+5: 3
+6: 5
+7: 8
+8: 13
+9: 21
+10: 34
+11: 55
+12: 89
+13: 144
+14: 233
+15: 377
+```
+
+And that's it!  You can augment `Program.cs` any way you like.
+
+## Adding some new files
+
+Single files are fine for simple one-off programs, but chances are you're going to want to break things out into multiple files if you're building anything which has multiple components.  Multiple files are a way to do that.
```

Comment:
Actually, there is another, better reason: reusable logic that you want in a library, so you begin a library by removing the said code into a separate file.